### DISCUSSION TITLE
Running MultiApps run "out of sync" vs Restart

### DISCUSSION THREAD

**Discussion by callaway-andrew-dykhuis** [idaholab/moose#28602] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi all, I have a problem setup question dealing with MultiApps and Restarts.

At a high level my simulation is this: heat transfer between block 1 and block 2 for a period of time, and then block 1 is removed and then heat transfer to/from block 2 via convection for a different period of time.

I have a working simulation where I use the Controls system to turn off a Dirichlet BC and turn on a convective BC. I have a larger simulation, with additional blocks, where I use ThermalContact to set up GapHeatTransfer objects for the temperature BC instead (more accurate heat transfer modeling than a DirichletBC based on a function, as in the first case). I'd prefer to just toggle these off and switch to a convective BC at a given point in time, but GapHeatTransfer is not controllable....and then I also have extra blocks that aren't involved in my physics at all.

From my review of the forum/discussions, it seems I have a few options:
- replace the GapHeatTransfer BC with a SideSetHeatTransferKernel, which I _believe_ is controllable. From my understanding, however, this would require me to create duplicate variables on each side of the interface, which I'd rather not do, and has the same extra mesh block issue as in my next point. 
- MultiApps: I think this might be the more straightforward option. Use GapHeatTransfer in the parent app, and then run a TransientMultiApp that executes on "final." The issue I'm having with this is getting the subapp to run more than a single timestep. I'm not sure how to use global_time_offset to set it up so that I run for say, 100 seconds, in the parent (from time = 0 to time = 100), and then 200 seconds (from time = 100 to time = 300) in the subapp AFTER the parent app in overall time. At that point the entire simulation should be done, even with the time mismatch, as I care about the subapp's final results. The mesh for the second portion has fewer blocks than the first portion of the problem, so setting this up as a separate simulation to me makes more sense than including extra mesh blocks that don't affect the heat transfer I care about.
- Restart - I haven't played around with this yet, but I'd be able to have the problem change from GapHeatTransfer to convective heat transfer by using the first output to initialize the second. Like the subapp input file, I could remove the extra blocks.

Additional information: 
- in the subapp input file I delete the extra mesh blocks using BlockDeletionGenerator, and therefore the mesh that remains is the same as the corresponding block in the parent app, but not close enough to MOOSE to allow use of MultiAppCopyTransfer. It is a smaller problem, however, so an option that lets me skip solving "block 1" and just focusing on block 2 is more efficient.
- I have multiple stateful properties. In the MultiApp approach I'd update initQpStatefulProperties to use a variable (if provided) as the initial condition, and this variable would be the transferred values from the parent (MaterialRealAux transferred to a dummy auxvar in the sub app, then coupled to the material). I think stateful properties are now supported in restarts, but perhaps not...or perhaps not in my specific use case here.
- MultiApp or Restart would also enable me to tack on additional steps...add back the original blocks and simulate another step, and repeat.

Any thoughts or guidance would be appreciated!

**Comment by GiudGiud** [comment]
Hello

I'm surpised we cant control `GapHeatTransfer`. 
But assuming that, there are a few options for the multiapps. 
- You can try to use offset 'start_time' of each app's transient executioners to get one to run before the other. With sub_cycling turned on, you should see one app (the one with the early start_time) catch up to the other one's start time. Then depending on the end_time specified, the first app might stop solving. 
- Easier way might be to just execute the multiapp and the transfers on FINAL. That way the main app runs to completion, then the sub-app is executed. 

For restart, we don't support changing the mesh (especially with different variable numbers of DOFs) if you use Checkpoint restarts. And we do not support restarting stateful properties if you use SolutionUO or Exodus restart. 
BUT
you could use a multiapp that restarts from the checkpoint with an identical mesh from the first simulation, then transfers to another app (either the main one or a subapp) that has the different mesh

>  but not close enough to MOOSE to allow use of MultiAppCopyTransfer

Use MultiAppGeneralFieldNearestLocationTransfer or ShapeEvaluation if the meshes strictly overlap

> MultiApp or Restart would also enable me to tack on additional steps...add back the original blocks and simulate another step, and repeat.

Multiapp probably best if you plan to do very many steps / iterations between the two simulations

**Reply by callaway-andrew-dykhuis** [reply]
Thanks for the quick response! Okay, so based on SolutionUO or Exodus restart I wouldn't have restart capability so MultiApp is the route. The thing is - I don't need the subapp to catch up to the parent: I want to run it like two separate simulations, just using the results from the end of the parent (execute_on = final) to start the subapp, which will then run past the end_time of the parent. 

Right now, I have the following settings (short times for debugging):
parent start_time = 0
parent end_time = 5.0
subapp start_time = 0
subapp end_time = 50.0
The subapp runs after the parent reaches 5.0 seconds, but it just runs a single timestep with dt equal to the final dt of the parent. Then it stops.

**Reply by GiudGiud** [reply]
you are using a TransientMultiApp though. You could use a FullSolveMultiApp here since you run a full calculation

**Reply by GiudGiud** [reply]
with a transientMultiApp you could do:
- execute_on = FINAL
- subapp start_time = -50
- subapp end_time = 0
that way, 'catching up' is running a full simulation for the subapp

**Reply by callaway-andrew-dykhuis** [reply]
Looks like FullSolveMultiApp takes care of it.

I just tried 0-5 for the parent and -50 to 0 for the subapp with a TransientMultiApp and I still get the one time step before it stops, but I believe FullSolveMultiApp solves my main issue. The stateful material initialization in the subapp should be doable via an auxvariable. I might lose a bit of resolution but I can live with that.

**Reply by GiudGiud** [reply]
> I just tried 0-5 for the parent and -50 to 0 for the subapp with a TransientMultiApp and I still get the one time step before it stop

Did you turn on sub_cycling? SHould have done that.
If not then catch_up = true is similar

**Reply by callaway-andrew-dykhuis** [reply]
The timing works out with FullSolveMultiApp, and using the adjusted start_time and end_time, but one additional question I have is in regards to initializing my stateful materials. I have optional coupling to an auxvariable set up so that if the variable is provided then that variable provides the value used in initQpStatefulProperties(). This works for a non-multiapp simulation to set the initial value of the material. However, in my multiapp I transfer material properties via MaterialRealAux to an auxvariable in the subapp, and then couple this auxvariable to the material, but I do not get the coupled value, just the default value. I assume it has something to do with when the transfer occurs vs when the material property is initialized? Is there an option to "re-initialize" the stateful material property?

**Reply by GiudGiud** [reply]
I dont think we have this option. 
Could you try this material?
https://mooseframework.inl.gov/source/materials/InterpolatedStatefulMaterial.html

**Reply by callaway-andrew-dykhuis** [reply, marked as answer]
I have something working for now, so I should be good. Thanks again! Based on some optional parameters I can couple to a variable for the _old value and then it functions as desired from there. 

For the record/to clean it up to mark this as the answer:
The fix was using FullSolveMultiApp that executes on "final," and where the subapp uses offset start_time and end_time values. This lets me run from say 0 to 100 seconds in the parent, and then 100 to 200 seconds in the subapp. I transfer material properties to the subapp using MultiAppGeneralFieldNearestLocationTransfer from the parent app's MaterialRealAux auxvariables to auxvariables in the subapp. I adjusted my stateful materials so that I could effectively use the transferred auxvariable in the initial properties calculation.
